src/prng.c

```diff
@@ -0,0 +1,93 @@
+/******************************************************************************
+ *
+ * Copyright 2017 Xaptum, Inc.
+ * 
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+#include <ecdaa/prng.h>
+
+#ifdef USE_LIBSODIUM_RNG_SEED_FUNCTION
+
+#include <sodium.h>
+#include <stdio.h>
+#include <stdlib.h>
+#if defined(__linux__)
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <linux/random.h>
+#include <stdio.h>
+#include <stdlib.h>
+#endif  // defined(__linux__)
+
+static ecdaa_prng_seed_function_type seed_func = &randombytes_buf;
+static int sodium_initialized = 0;
+
+static void sodium_init_if_needed(void);
+
+#else   // USE_LIBSODIUM_RNG_SEED_FUNCTION
+
+static ecdaa_prng_seed_function_type seed_func;
+
+#endif   // USE_LIBSODIUM_RNG_SEED_FUNCTION
+
+void ecdaa_prng_init(struct ecdaa_prng *prng_in)
+{
+#ifdef USE_LIBSODIUM_RNG_SEED_FUNCTION
+    sodium_init_if_needed();
+#endif
+
+	RAND_clean(&prng_in->impl);
+
+    char seed[128]; // 128 as recommended in AMCL's "rand.c"
+
+    seed_func(seed, sizeof(seed));
+
+    RAND_seed(&prng_in->impl, sizeof(seed), seed);
+}
+
+void ecdaa_prng_set_seed_function(ecdaa_prng_seed_function_type new_function)
+{
+    seed_func = new_function;
+}
+
+#ifdef USE_LIBSODIUM_RNG_SEED_FUNCTION
+void sodium_init_if_needed(void)
+{
+    if (1 != sodium_initialized) {
+        if (-1 == sodium_init()) {
+            fputs("Unable to initialize libsodium random seed source. Aborting", stderr);
+            abort();
+        }
+#if defined(__linux__) && defined(RNDGETENTCNT)
```

Comment:
I don't think this section belongs in a low-level crypto library - for the same reasons that it's not included in `libsodium` by default.

How the program should respond to low entropy is application-specific.  For many services, the right answer might be to keep running, but refuse connections until sufficient entropy is available.

Reply:
May be better to let the `ecdaa_init()` have the same semantics as the `sodium_init()` return code. Then copy the that code snippet from the `libsodium` readme to the `ecdaa` readme.